Inline reply functionality

irgendwie opened the pull request:
It's work in progress, but should already work for text messages.
Edit: PR for #648 @jancborchardt

jancborchardt commented on js/mail.js:
```diff
@@ -627,6 +627,17 @@ var Mail = {
+							data.replyBody = '> ' + jQuery(text).text().replace(/\n/g, '\n> ');
+						});*/
+					} else {
+						data.replyBody = '\n\nOn ' + data.date + ' ' + data.from + ' wrote:\n> ' + jQuery(data.body).text().replace(/\n/g, '\n> ');
```
We need to make this translatable. But I’m actually thinking if it’s better in a simpler format. Instead of:
* On 17 mai 2015 14:40:18 UTC+2 Jan-Christoph Borchardt wrote:
Better
* Jan-Christoph Borchardt (17 May 2015, 14:40)

Like a chat format.

DeepDiver1975 replied:
gmail format:

````
rftgzhujioko

2015-05-10 14:23 GMT+02:00 admin <deepdiver1975@gmail.com>:
dfsfdsfasdasd

DeepDiver1975 replied:
In terms of international usage we shall keep the timezone?

jancborchardt commented:
Nice! :) Just the few things:

* there should be two (or even three) blank lines above the quote instead of just one
* check my comment about the different format for the quote header
* yeah, we have to make it work for HTML mails. Otherwise it’s super-inconsistent for which mails it works. Especially as we format most simple HTML mails to look like text mails.

Great work @irgendwie! :) Let’s continue with it on Wednesday.

jancborchardt commented:
Btw @irgendwie one thing to improve would be more descriptive branch names. :) For example, `inline-reply` instead of `fix-648`.

irgendwie commented:
@jancborchardt I'll have that in mind for the future ;-)

DeepDiver1975 commented:
> Btw @irgendwie one thing to improve would be more descriptive branch names. :) For example, inline-reply instead of fix-648.

fix-648 is highly descriptive to me

DeepDiver1975 commented:
> yeah, we have to make it work for HTML mails. Otherwise it’s super-inconsistent for which mails it works. Especially as we format most simple HTML mails to look like text mails.

we need to dig into the html purifyer to see what we can do - maybe @LukasReschke can help out tomorrow as we meet :rocket:

jancborchardt commented:
> fix-648 is highly descriptive to me

You confuse »descriptive« with »accurate«. ;) The issue can be mentioned in the commit message. When you do a git branch to show all branches and only see a list of fix-* branches, you have no clue what they are about.

DeepDiver1975 commented:
> you have no clue what they are about.

to be honest - it's quiet simple - they fix that issue .... :wink:

DeepDiver1975 commented:
@irgendwie there are some js issue - see travis

````
> jshint . && ./node_modules/.bin/jsxhint .

js/mail.js: line 633, col 29, Bad line breaking before '+'.

js/mail.js: line 634, col 29, Bad line breaking before '+'.

js/mail.js: line 635, col 29, Bad line breaking before '+'.

js/mail.js: line 683, col 29, Bad line breaking before '+'.

js/mail.js: line 684, col 29, Bad line breaking before '+'.

js/mail.js: line 685, col 29, Bad line breaking before '+'.

js/mail.js: line 634, col 31, 'moment' is not defined.

js/mail.js: line 684, col 31, 'moment' is not defined.

irgendwie commented:
@jancborchardt Review please! HTML handling should be much better now!

jancborchardt commented:
This is pretty much awesome now!

Just one little nitpick with the date format (which you will be able to handle in no time now you are a master of moment.js ;):
Leading zeroes in the day of the month and in the hour should be removed. That is »05. May« should be »5. May«. And »04:20« should be »4:20«.

Otherwise :+1: :) (I also checked off the done items in my comment above.)

jancborchardt commented on js/mail.js:
```diff
@@ -671,6 +678,14 @@ var Mail = {
+					document.getElementsByClassName('reply-message-body')[0].innerText = (
+						'\n\n\n' +
+						data.from + ' – ' +
+						moment(data.dateIso).format('ddd, DD. MMM YYYY HH:mm') + '\n> ' +
```
I suppose the comment I made about the date format would be solved here by using `ddd, D. MMM YYYY H:mm` instead?

irgendwie commented:
:imp: those... hours... wasted :D
fixed!

jancborchardt commented:
:+1: I for one welcome this PR as our new overlord.

jancborchardt commented on js/mail.js:
```diff
@@ -631,6 +631,14 @@ var Mail = {
 					data.replyBody = draft.body;
+				} else {
+					// Add body content to inline reply (text mails)
+					if(!data.hasHtmlBody) {
```
Yep. I also thought before that `if(` was better style but the JS guidelines from both Mozilla and Google say `if (`. :)

jancborchardt commented:
@DeepDiver1975 @irgendwie I assume the JS issues mentioned above by @DeepDiver1975 are also fixed? https://github.com/owncloud/mail/pull/679#issuecomment-104182353

wurstchristoph commented on js/mail.js:
```diff
@@ -675,6 +683,16 @@ var Mail = {
 					// Remove spinner when loading finished
 					$('iframe').parent().removeClass('icon-loading');
+					// Add body content to inline reply (html mails)
+					var text = $(this).contents().find('body')[0].innerText;
```
I get an error in FF saying ``TypeError: text is undefined``

I fixed it by using ``var text = $(this).contents().find('body').first().text();``

wurstchristoph commented on js/mail.js:
```diff
@@ -675,6 +683,16 @@ var Mail = {
+					// Add body content to inline reply (html mails)
+					var text = $(this).contents().find('body')[0].innerText;
+					if(!draft) {
+						document.getElementsByClassName('reply-message-body')[0].innerText = (
```
This can be replaced by jQuery like ``$('.reply-message-body').first().text(...);`` I guess

jancborchardt commented:
@wurstchristoph do you have anything against merging this and doing fixes in follow-up PRs? As far as I tested nothing is broken by this PR – we can polish incrementally and shouldn’t let this linger for too long. :)

irgendwie commented:
`js/mail.js: line 639, col 29, 'moment' is not defined.`
`js/mail.js: line 692, col 29, 'moment' is not defined.`

@DeepDiver1975  any idea on that?

wurstchristoph commented:
inline replies work very well unless the original message is HTML:
![](https://cloud.githubusercontent.com/assets/1374172/7757573/bbcba962-0003-11e5-9a3f-aa22845c2c45.png)

As commented above, using ``var text = $(this).contents().find('body').text();`` fixes the issue for me :confused:

I tested with FF with a github mail and a HTML mail composed with evolution

@jancborchardt I would say this needs to be fixed before it is merged

jancborchardt commented:
Yeah, we need to fix Travis and the JS there.

@irgendwie does @wurstchristoph’s fix work for you? @biolumin in case you’re up for helping with the JS.

irgendwie commented:
@wurstchristoph added your suggestions, would you mind reviewing it? :-)
cc @jancborchardt and @DeepDiver1975 (for the `'moment' is not defined`)

wurstchristoph commented:
@irgendwie the JS error is fixed :-)

However, I found another bug: the textarea element is not resized automatically, I need to click into it and press some key (e.g. space), then the element is resized to show the full content:
# Before
![](https://cloud.githubusercontent.com/assets/1374172/7774027/8aeb750c-00a8-11e5-8acc-8d136a6f6410.png)
# After pressing a key inside the textarea
![](https://cloud.githubusercontent.com/assets/1374172/7774028/8af04ea6-00a8-11e5-8171-1f953de79cca.png)

It seems that this happens if the content of the textarea is changed after the textarea was created... Maybe jQuery autosize has to be triggered after loading the reply content into the textarea. Maybe https://github.com/owncloud/mail/blob/b7977ae2dbd51c7cda2ea32280538d6e96e08794/js/views/sendmail.js#L149 helps

irgendwie commented:
@wurstchristoph this is intended - the reply message should be default size at the beginning, so that it is clear, what its purpose is. If the reply message body is extended from the beginning, maybe it isn't clear what it is used for unless you completely scroll down until you see the reply button. It's more or less a feature request of @jancborchardt and I think this sounds reasonable.

I already implemented a resize on click event, so that if you click in the reply message body, the text area is expended. Still, drafts and text messages are completely expanded from the beginning - this needs to be "fixed".

wurstchristoph commented:
:+1: looks good to me

irgendwie commented:
@jancborchardt I implemented the size/resize as discussed at the meetup!
Additionally, I added a resize on click event, to expand the reply message body, which I think is an intuitive behavior.
So this last change should solve your last checkbox:
> reply box should only expand once you edit the content. Otherwise the reply button is not visible and it might be confusing.

Some code isn't pretty, but functional :see_no_evil:

jancborchardt commented:
Nice, good stuff, also on the resize! :) Am mobile atm - what do we do with the momentjs stuff for ownCloud 7 compat? Just catch it and use another simple date function? We shouldn't ship it separately I guess.

irgendwie commented:
Does this work for everyone?
@jancborchardt @wurstchristoph

wurstchristoph commented:
@irgendwie great that you could remove the moment.js dependency, works very well :+1:

jancborchardt commented:
Looks very nice, let's finally get this in! :)
